go/tasks/plugins/array/core/state.go

```diff
@@ -267,8 +263,9 @@ func SummaryToPhase(ctx context.Context, minSuccesses int64, summary arraystatus
 		return PhaseWriteToDiscoveryThenFail
 	}
 
-	// No chance to reach the required success numbers.
-	if totalRunning+totalSuccesses+totalWaitingForResources+totalRetryableFailures < minSuccesses {
+	// No chance to reach the required success numbers. If any subtasks are running we let them
+	// complete, so results may be written to cache if enabled, before failing.
+	if totalSuccesses+totalWaitingForResources+totalRetryableFailures < minSuccesses && totalRunning == 0 {
```

Comment:
Should we just add total running to the +s

Reply:
OK, so I felt strongly enough to change it, but not enough to defend the change. I reverted this to the previous approach. Basically, I think transitioning to permanent failure should operate differently whether the task is cached or not:
Cached: execute as many of the subtasks as possible to allow the cached outputs to be written.
Not cached: fail as soon as possible so you don't spend any time doing unnecessary work (current approach).